### PR TITLE
Add LLVM API support for address spaces

### LINKED ISSUE #188: Add LLVM API support for address spaces
Needs

* pointers with address spaces,
* addrspacecast

Needed for #199

### PR DESCRIPTION
**Pull request by heshanpadmasiri**
## Purpose

- `PointerType` with address spaces
-  `addrspacecast` support

Resolves #188

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: compiler/modules/print.llvm/tests/testOutputs/addr_space_cast.ll [hunk 1]
```diff
@@ -0,0 +1,6 @@
+declare i64 addrspace (3 )* @genFn ()
+define i64* @main () {
+  %_0 = call i64 addrspace (3 )* @genFn ()
+  %_1 = addrspacecast i64 addrspace (3 )* %_0 to i64*
```
Comment (KavinduZoysa) [inline compiler/modules/print.llvm/tests/testOutputs/addr_space_cast.ll]: Can we remove this extra space here? 
`addrspace (3 )*` -> `addrspace (3)*`

File: compiler/modules/print.llvm/llvm.bal [hunk 2]
```diff
@@ -650,7 +671,11 @@ function sameIntType(Value v1, Value v2) returns IntType {
 function typeToString(RetType ty) returns string {
     string typeTag;
     if ty is PointerType {
-        typeTag = typeToString(ty.pointsTo) + "*";
+        if ty.addressSpace == 0 {
+            typeTag = typeToString(ty.pointsTo) + "*";
+        } else {
+            typeTag = createLine([typeToString(ty.pointsTo), "addrspace", "(" , ty.addressSpace.toString(), ")*"]);
```
Comment (jclark) [inline compiler/modules/print.llvm/llvm.bal]: I think you need to split `")*"` into `")", "*"` so `omitSpaceBefore` can leave out the extra space (which @KavinduZoysa pointed out)

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/print.llvm/llvm.bal (hunk 2 above)
Comment (heshanpadmasiri) [reply compiler/modules/print.llvm/llvm.bal]: Fixed

File: compiler/modules/print.llvm/tests/testOutputs/addr_space_cast.ll (hunk 1 above)
Comment (heshanpadmasiri) [reply compiler/modules/print.llvm/tests/testOutputs/addr_space_cast.ll]: Fixed

**Pull request merged**